### PR TITLE
serializers: Get referenced records from ES

### PR DESCRIPTION
**Pull request by pazembrz**
* INSPIR-2713

### CODE REVIEW

File: backend/inspirehep/records/api/literature.py [hunk 1]
```diff
@@ -89,6 +89,45 @@ def create(
...
+            >>> data = {
+                'references': [
+                    {
+                        'record': {
+                            '$ref': 'http://localhost/literature/1'
+                        }
+                    }
+                ]
+            }
+            >>>  records = LiteratureRecord.get_linked_records_from_field(
```
Comment (ammirate) [inline backend/inspirehep/records/api/literature.py]: I love examples like this, it is missing the function output though ;)

File: backend/inspirehep/records/api/literature.py [hunk 2]
```diff
@@ -89,6 +89,45 @@ def create(
...
+        """
+        pids = cls._get_linked_pids_from_field(data, path)
+        if pids:
+            results = LiteratureSearch.get_records_by_pids(pids)
+            for result in results:
+                try:
+                    rec_data = json.loads(result._ui_display)
+                except AttributeError:
+                    LOGGER.exception(
+                        "Record do not have _ui_display field!", record=result.meta.id
```
Comment (ammirate) [inline backend/inspirehep/records/api/literature.py]: ops, `doesn't`.
Also, I think we have a lot of logs with recid reather than records uuid

Comment (pazembrz) [reply backend/inspirehep/records/api/literature.py]: UUID here is the only thing which will be there always. recid might be missing from there if the record won't be indexed properly (and it's not indexed properly if _ui_display field is missing).

File: backend/inspirehep/records/api/literature.py [hunk 3]
```diff
@@ -89,6 +89,53 @@ def create(
...
+                    )
+                    continue
+                result._d_.update(rec_data)
+                yield InspireRecord.get_class_for_record(result._d_)(result._d_)
```
Comment (michamos) [inline backend/inspirehep/records/api/literature.py]: * Isn't this always a `LiteratureRecord` given you're doing a `LiteratureSearch`?
* I know the semantics of this in invenio are not very clear (you can have a `Record` that's not backed by a model instance) but why should this `LiteratureRecord`?

File: backend/inspirehep/records/api/literature.py [hunk 4]
```diff
@@ -89,6 +89,53 @@ def create(
...
+        """
+        pids = cls._get_linked_pids_from_field(data, path)
+        if pids:
+            results = LiteratureSearch.get_records_by_pids(pids)
+            for result in results:
+                try:
+                    rec_data = json.loads(result._ui_display)
+                except AttributeError:
+                    LOGGER.exception(
+                        "Record do not have _ui_display field!", record=result.meta.id
```
Comment (michamos) [inline backend/inspirehep/records/api/literature.py]:
```suggestion
                        "Record does not have _ui_display field!", record=result.meta.id
```

File: backend/inspirehep/records/api/literature.py [hunk 5]
```diff
@@ -89,6 +89,53 @@ def create(
...
+                    )
+                    continue
+                result._d_.update(rec_data)
```
Comment (michamos) [inline backend/inspirehep/records/api/literature.py]: what is this?

File: backend/inspirehep/records/api/literature.py [hunk 6]
```diff
@@ -89,6 +89,53 @@ def create(
...
+                    'collaborations': [(...)],
+                }
+            ]
+        """
+        pids = cls._get_linked_pids_from_field(data, path)
+        if pids:
+            results = LiteratureSearch.get_records_by_pids(pids)
+            for result in results:
+                try:
+                    rec_data = json.loads(result._ui_display)
```
Comment (michamos) [inline backend/inspirehep/records/api/literature.py]: how do you know the data you're interested in is actually present in `_ui_display`? and what's the point of using that if you're deserializing it anyway? wouldn't it make more sense to have a configurable `_source`?

File: backend/inspirehep/records/api/literature.py (hunk 5 above)
Comment (pazembrz) [reply backend/inspirehep/records/api/literature.py]: `_d_` is a dictionary of the record from ES, I'm updating it with data loaded from `_ui_display` json.

File: backend/inspirehep/records/api/literature.py (hunk 6 above)
Comment (pazembrz) [reply backend/inspirehep/records/api/literature.py]: We don't have everything in `_ui_display`, we are missing  `arxiv_eprints`, `collaborations`, `dois` and `misc` field from there, this is why we are merging deserialized `_ui_display` with this missing fields in line 135. I preffer to use `_ui_data` as this is only place where we have limited number of authors there. Big number of authors was the main reason why loading of references was so slow before.

File: backend/inspirehep/search/api.py [hunk 7]
```diff
@@ -145,6 +145,22 @@ def citations(record, page=1, size=10):
         )
         return citations_search.execute().hits
 
+    @staticmethod
+    def get_records_by_pids(pids, source=None, size=1000):
```
Comment (michamos) [inline backend/inspirehep/search/api.py]: what if there are more than 1000 refs?

Comment (pazembrz) [reply backend/inspirehep/search/api.py]: :+1:

**Pull request merged**